### DISCUSSION TITLE
BL_CARVEOUT: Failed to allocate memory of size 0x36800000

### DISCUSSION THREAD

**Discussion by gsally-sarcos** [OE4T/meta-tegra#1284] [Q&A]
Hi,
I'm adding a new board (Jetson Orin on ConnectTech Forge) Hopefully, we can get this into the project as community supported.

I think I followed along correctly adding the right flashvars and creating a new entry in conf/machine, using the Jetson Orin as the starting point and updating the SKU.  I swiped the device tree from the vendor's Nvidia-based SDK.  So far so good.   I can build and the when the programming tool runs, I see this on the console:

```
[   3.8503 ] Sending membct and RCM blob
[   3.8516 ] tegrarcm_v2 --chip 0x23 0 --pollbl --download bct_mem mem_rcm_sigheader.bct.encrypt --download blob blob.bin
[   3.8524 ] BL: version 0.23.0.0-t234-54845784-92215740 last_boot_error: 0
[   3.8809 ] Sending bct_mem
[   3.9239 ] Sending blob
[   3.9565 ] ERROR: might be timeout in USB write.
```

And this from the debug serial port (thankfully, this board has the debug port wired!)

```
[0025.820] I> Non-ECC region[3]: Start:0x0, End:0x0
[0025.825] I> Non-ECC region[4]: Start:0x0, End:0x0
[0025.835] E> BL_CARVEOUT: Failed to allocate memory of size 0x36800000 for CO:46.
[0025.842] C> Task 0x0 failed (err: 0x49490003)
[0025.847] E> Top caller module: BL_CARVEOUT, error module: BL_CARVEOUT, reason: 0x03, aux_info: 0x00
[0025.856] I> Busy Spin
```

The USB timeout seems to be the result of the error ``` BL_CARVEOUT: Failed to allocate memory of size 0x36800000```.  However, I'm having a difficult time tracking down what that number means and its origin.  ```tegra234-mb1-bct-carveout-l4t.dtsi```  appears to be the source (changing this value results in the changed value appearing in the error message)

```
                        aux_info@CARVEOUT_VPR {
                                pref_base = <0x0 0x0>;
                                size = <0x0 0x36800000>; // 872MB
                                alignment = <0x0 0x100000>; // 1MB
                        };
```

But from there, I can't track down the source.   It's not clear how the ```tegra234-mb1-bct-carveout-l4t.dtsi``` is generated (I've been looking) so any information there would be helpful as well.

**Comment by dwalkes** [comment]
I don't have any quick answers - however I did notice the same error in the thread at https://forums.developer.nvidia.com/t/flashing-nano-devkit-nvme-over-usb-fails/251189/84 and https://forums.developer.nvidia.com/t/orinnx-8gb-flash-support-for-customboard-in-35-3-1/251187/25

Have you tried bringing up with stock NVIDIA Jetpack yet?  This might be worth trying/comparing.

**Comment by madisongh** [comment]
Which Orin module are you using?  We don't currently have support for all of the different Orin flavors in the layer.

**Reply by gsally-sarcos** [reply]
Hi @madisongh, Using Orin AGX 64gb, running on connecttech forge board.    The jetpack based BSP from the vendor refuses to install (the installation script contained various syntax errors, fixed those and then ran into other problems, but that's another story...) 

I changed by build to run the orin agx for the devkit, figuring it would panic in some way at runtime and I could go from there to get things more or less working.  The orin AGX as it is installs through completion, but does not start the MB2 bootloader.  I see this on the debug console.  This is a better hint than I had before.  I attached the entire file if that provides any more clues, in the meanwhile, I'll feel through how to make use of the bct files included in the vendor's BSP in this build, maybe that will make a difference.  I wasn't worried too much about the flash layout, figuring that extra flash memory not described by flash.xml would just be unused. 

```
[0031.228] I> Loading MB2 Applet
[0031.231] I> Slot: 1
[0031.233] E> BLOCK_DEV: Failed to open blockdev.
[0031.238] E> LOADER: Failed to open blockdev 0(0).
[0031.243] E> LOADER: Failed to get storage info for binary 21 from loader.
[0031.250] C> LOADER: Could not read binary 21.
[0031.254] E> Failed to load MB2
[0031.257] C> Task 0x0 failed (err: 0x27228311)
[0031.261] E> Top caller module: MB2_PARAMS, error module: LOADER, reason: 0x11, aux_info: 0x83
[0031.270] C> Boot Info Table status dump :
1
```

[orin-64gb-m1-log.txt](https://github.com/OE4T/meta-tegra/files/11984778/orin-64gb-m1-log.txt)

**Reply by gsally-sarcos** [reply]
I did end-up making a change to the BCT file in the eeprom section.  What's the right way to get this into the meta-tegra project, in terms of having an overlay specific to the Orin-AGX/Forge module/carrier?  More changes will be necessary, so knowing the general steps will be helpful. 

```
        eeprom {
            cvm_eeprom_i2c_instance = <0>;
            cvm_eeprom_i2c_slave_address = <0xa0>;
            cvm_eeprom_read_size = <0x100>;
            cvb_eeprom_i2c_instance = <0x0>;
            cvb_eeprom_i2c_slave_address = <0xac>;
            cvb_eeprom_read_size = <0x0>;
        };
```

**Reply by madisongh** [reply]
Wow, so mb2 was choking on there being no CVB EEPROM?

> What's the right way to get this into the meta-tegra project, in terms of having an overlay specific to the Orin-AGX/Forge module/carrier?

The answer probably depends on what sort of other changes are needed.  Some of the carrier providers have made extensive kernel modifications - adding and modifying drivers, making changes to the NVIDIA device tree files, etc. - that would make direct integration (and ongoing maintenance) painful.  If Connecttech have done that for this carrier, it might make sense to have the machine configs, boot files, and a separate kernel recipe for their custom kernel, in another layer.

On the other hand, if the changes are minimal, we could add a machine config, tegra-flashvars file, and recipes for providing carrier-specific boot config files (or modifying the stock ones for that particular machine, as needed) here in meta-tegra directly.

**Reply by gsally-sarcos** [reply]
I haven't dug into all of the changes for this board.  The pinmux is different (that's to be expected), so I updated the flashvars to use the pixmux device tree from the vendor.  I'll have to see what was done in terms of kernel configuration, etc.  The machine boots the kernel, mounts the RFS and I can get to the serial prompt over the debug port.  Networking isn't functioning (the PHYS look to be in reset...) but I'm on the right track.    I'll just keep moving forward one step at a time. :)

The `./doflash.sh` is outputting this message, not sure if that's expected output, as the programming works and the kernel starts

```
[   4.1492 ] MB1-BCT version: 0.10
[   4.1493 ] ERROR: carveout /misc/carveout/aux_info@CARVEOUT_UNUSED5/ is not supported
[   4.1495 ] ERROR: carveout /misc/carveout/aux_info@CARVEOUT_UNUSED5/ is not supported
[   4.1497 ] ERROR: carveout /misc/carveout/aux_info@CARVEOUT_UNUSED5/ is not supported
[   4.1498 ] ERROR: carveout /misc/carveout/aux_info@CARVEOUT_MCE_COVERAGE/ is not supported
[   4.1500 ] ERROR: carveout /misc/carveout/aux_info@CARVEOUT_MCE_COVERAGE/ is not supported
[   4.1501 ] ERROR: carveout /misc/carveout/aux_info@CARVEOUT_MCE_COVERAGE/ is not supported
[   4.1503 ] ERROR: /misc/tsc_controls/tsc_locking_config is not supported
[   4.1504 ] ERROR: /misc/tsc_controls/tsc_locking_diff_configuration is not supported
[   4.1505 ] ERROR: /misc/tsc_controls/tsc_locking_ref_frequency_configuration is not supported
[   4.1507 ] ERROR: /misc/tsc_controls/tsc_locking_control is not supported
[   4.1508 ] ERROR: /misc/tsc_controls/tsc_locking_adjust_configuration is not supported
[   4.1510 ] ERROR: /misc/tsc_controls/tsc_locking_fast_adjust_configuration is not supported
[   4.1511 ] ERROR: /misc/tsc_controls/tsc_locking_adjust_delta_control is not supported
[   4.1513 ] ERROR: /misc/tsc_controls/tsc_capture_control_ptx is not supported
[   4.1514 ] ERROR: /misc/tsc_controls/tsc_capture_config_ptx is not supported
[   4.1515 ] ERROR: /misc/tsc_controls/tsc_stscrsr is not supported
[   4.1516 ] ERROR: /misc/tsc_controls/tsc_locking_adjust_num_control is not supported
```

**Reply by madisongh** [reply]
I don't see those errors on my AGX Orin dev kit,and I also don't see a message reporting the MB1-BCT version.  So I'm not sure what's going on there.
